python/python/glide/async_commands/core.py

```diff
@@ -1368,3 +1373,111 @@ async def invoke_script(
                 ["foo", "bar"]
         """
         return await self._execute_script(script.get_hash(), keys, args)
+
+    @dataclass
+    class PubSubMsg:
+        """Describes incoming message
+
+        Attributes:
+            message (str): Incoming message.
+            channel (str): Name of an channel that triggered the message.
+            pattern (Optional[str]): Pattern that triggered the message.
+            context (Optional[Any]): User-provided context for this subscription.
+        """
+
+        message: str
+        channel: str
+        pattern: Optional[str]
+        context: Optional[Any]
+
+    class ChannelModes(Enum):
+        """Describes subsciption modes."""
+
+        Exact = 1
+        """ Use exact channel names """
+        Pattern = 2
+        """ Use channel name patterns """
+        Sharded = 3
+        """ Use sharded PUBSUB. See https://redis.io/docs/interact/pubsub/#sharded-pubsub for more details """
+
+    async def subscribe(
+        self,
+        channels_or_patterns: Set[str],
+        channel_mode: CoreCommands.ChannelModes,
+        callback: Callable[[CoreCommands.PubSubMsg], None],
+        context: Optional[Any],
+    ) -> None:
+        """
+        Add specific channels or channel patterns to active subscriptions.
+        The callback is associated with each channel/pattern,
+        meaning subsequent subsciption for the same channel/pattern will override the previous association,
+        dereferencing the original callback object.
+        Note1: Overlapping subscriptions produced by combinations of Exact and Pattern modes
+        count as distinct subscriptions and will produce duplicated messages.
+        Note2: Patterns are not applicable in Sharded mode. That is, it is not possible to create a sharded subscription with a pattern,
+        and pattern/exact subscriptions won't receive messages which were published in Sharded mode.
+        See https://redis.io/docs/interact/pubsub for more details.
+
+        Args:
+            channels_or_patterns (Set[str]): Set of channels or patterns to subscibe to.
+            channel_mode (ChannelModes): Mode of operation. See ChannelModes.
+            callback (Callable[[CoreCommands.PubSubMsg], None]): Callback to be called for each new message.
+            For each channel or pattern there is exactly one active callback.
+                    Subsequent calls to subscribe() will reset callbacks for existing channels or patterns.
+            context (Optional[Any]): User-provided context for this subscription. Will be passed to callback as part of PubSubMsg.
+
+        Examples:
+            >>> def process_news_message(message: PubSubMsg) -> None:
+            >>>     print(f"'{message.context}' received new message '{message.message}' from channel '{message.channel}')
```

Comment:
It isn't clear from the documentation that the context will be provided as a member of PubSubMsg, without the example I wouldn't know this is how it should be used.
Why not providing the context as an argument of the callback function?

Reply:
```
def process_news_message(message: PubSubMsg, context: Optional[Any]) -> None:
   print(f"'{context}' received new message '{message.message}' from channel '{message.channel}')

Reply:
I would do this more explanatory example:
```
def process_new_message(message: PubSubMsg, context: Dict[str, List[str]]):
    context.set_default(message.channel, [])
    context[message.channel].append(message.message)